lib/src/main/java/com/diffplug/spotless/PaddedCell.java

```diff
@@ -171,4 +173,82 @@ public String userMessage() {
 		}
 		// @formatter:on
 	}
+
+	/**
+	 * Calculates whether the given file is dirty according to a PaddedCell invocation of the given formatter.
+	 * DirtyState includes the clean state of the file, as well as a warning if we were not able to apply the formatter
+	 * due to diverging idempotence.
+	 */
+	public static DirtyState calculateDirtyState(Formatter formatter, File file) throws IOException {
+		Objects.requireNonNull(formatter, "formatter");
+		Objects.requireNonNull(file, "file");
+
+		byte[] rawBytes = Files.readAllBytes(file.toPath());
+		String raw = new String(rawBytes, formatter.getEncoding());
+		String rawUnix = LineEnding.toUnix(raw);
+
+		// enforce the format
+		String formattedUnix = formatter.compute(rawUnix, file);
+		// convert the line endings if necessary
+		String formatted = formatter.computeLineEndings(formattedUnix, file);
+
+		// if F(input) == input, then the formatter is well-behaving and the input is clean
+		byte[] formattedBytes = formatted.getBytes(formatter.getEncoding());
+		if (Arrays.equals(rawBytes, formattedBytes)) {
+			return isClean;
+		}
+
+		// F(input) != input, so we'll do a padded check
+		String doubleFormattedUnix = formatter.compute(formattedUnix, file);
+		if (doubleFormattedUnix.equals(formattedUnix)) {
+			// most dirty files are idempotent-dirty, so this is a quick-short circuit for that common case
+			return new DirtyState(formattedBytes);
+		}
+
+		PaddedCell cell = PaddedCell.check(formatter, file, rawUnix);
+		if (!cell.isResolvable()) {
+			return didNotConverge;
+		}
+
+		// get the canonical bytes
+		String canonicalUnix = cell.canonical();
+		String canonical = formatter.computeLineEndings(canonicalUnix, file);
+		byte[] canonicalBytes = canonical.getBytes(formatter.getEncoding());
+		if (!Arrays.equals(rawBytes, canonicalBytes)) {
+			// and write them to disk if needed
+			return new DirtyState(canonicalBytes);
+		} else {
+			return isClean;
+		}
+	}
+
+	/**
+	 * The clean/dirty state of a single file.  Intended use:
+	 * - {@link #isClean()} means that the file is is clean, and there's nothing else to say
+	 * - {@link #isConverged()} means that we were able to determine a clean state
+	 * - once you've tested the above conditions and you know that it's a dirty file with a converged state,
+	 *   then you can call {@link #canonicalBytes()} to get the canonical form of the given file.
+	 */
+	public static class DirtyState {
+		private final byte[] canonicalBytes;
+
+		private DirtyState(byte[] canonicalBytes) {
+			this.canonicalBytes = canonicalBytes;
+		}
+
+		public boolean isClean() {
+			return this == isClean;
+		}
+
+		public boolean didNotConverge() {
+			return this == didNotConverge;
+		}
+
+		public byte[] canonicalBytes() {
+			return Objects.requireNonNull(canonicalBytes, "First make sure that `!isClean()` and `!didNotConverge()`");
+		}
```

Comment:
I don't know if DiffPlug as an equivalent/copy of Guava's `Bytes` class under the `com.diffplug.common` package, but if we do, have you considered returning `Collections.unmodifiableList(Bytes.asList(canonicalBytes))` to ensure that non-Spotless users of `spotless-lib` can't mutate the underlying array?

Alternatively, have you considered copying `canonicalBytes` when it comes in through the `DirtyState` constructor and copying it out again from `canonicalBytes()`; or just documenting here that the user shouldn't mutate the array?

(If Guava had an `ImmutableByteArray`, I'd recommend updating `com.diffplug.common` and using that, but alas they only have array wrappers for `int`, `long` and `double`.)

Reply:
I've just seen where `DirtyState::canonicalBytes` is being used, and given that it's being passed to `Files::write`, I think the option of returning an unmodifiable list is probably the most expensive one of the lot, since `toArray(new byte[0])` would need to be called and lots of temporary `Byte` objects would be created.

If it were me, I'd copy `canonicalBytes` in and out of the `DirtyState` object, but I'm open to just adding a short Javadoc saying that the user shouldn't modify the returned array. :)

Reply:
Ah, it occurs to me that `DirtyState`'s constructor is private, so as an optimization for the copying option, we could just copy the array out of the `DirtyState` object; no need to copy it as it comes into the constructor.

Reply:
How about if I replace `byte[] canonicalBytes()` with `void writeCanonicalTo(File)` and `void writeCanonicalTo(OutputStream out)`.  That protects the `byte[]` from user corruption without any performance penalty.

Reply:
@nedtwigg That sounds like a good change to me! Go right ahead. 👍

(You might have guessed that I prefer `Path` over `File`. If it were me, `Path` is what I'd use, but given that `File` is still used pervasively by Gradle, I'm happy either way. 😄)